app/controllers/password_resets_controller.rb

```diff
@@ -0,0 +1,65 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: [:edit, :update]
+  before_action :valid_user, only: [:edit, :update]
+  before_action :check_expiration, only: [:edit, :update]
+
+  def new; end
+
+  def create
+    @user = User.find_by email:
+      params[:password_reset][:email].downcase
+    if @user
+      @user.create_reset_digest
+      @user.send_password_reset_email
+      flash[:info] = t "reset_pass.info"
+      redirect_to root_url
+    else
+      flash.now[:danger] = t "reset_pass.danger"
+      render :new
+    end
+  end
+
+  def update
+    if user_params[:password].blank?
+      flash.now[:fail] =
+        @user.errors.add(:password, t( "reset_pass.blank" ))
+      render :edit
+    elsif @user.update_attributes user_params
+      log_in @user
+      flash[:success] = t "reset_pass.success"
+      redirect_to @user
+    else
+      render :edit
+    end
+  end
+
+  def edit; end
+
+  private
+
+  def get_user
+    @user = User.find_by email: params[:email]
+    unless @user
+      flash[:warning] = t "reset_pass.warning"
+      redirect_to root_url
+    end
+  end
+
+  def user_params
+    params.require(:user).permit :password, :password_confirmation
+  end
+
+  def valid_user
+    authenticated = @user.authenticated?(:reset, params[:id])
```

Comment:
Theo request thì cái params[:id] này sẽ lưu giá trị gì e? 
Trong DB mình lưu thêm cái trường `reset_digest` để làm gì e nhỉ ?

Reply:
Theo em hiểu thì cái params[:id] lưu cái token mà mình bỏ vào trong cái link để reset pass.
Trong DB lưu cái reset_digest để so sánh với cái token ở trên á anh.
Khi submit email để reset mật khẩu sau khi check các kiểu thì sẽ tạo 1 cái new token rồi update field reset_digest, sau đó truyền thuộc tính reset_token của @user vào link reset, người dùng nhấn vào link thì nó sẽ so sánh cái token trên request vs cái reset_digest trong db để đảm bảo tính an toàn cho request, theo em hiểu là vậy á (@@).